python/spinn/fat_stack.py

```diff
@@ -373,6 +384,35 @@ def run(self, print_transitions=False, run_internal_parser=False,
         return [stack[-1] for stack in self.stacks], transition_loss
 
 
+    def reinforce(self, rewards):
+        statistics = zip(*[
+            (m["hyp_acc"], m["truth_acc"], m["hyp_xent"], m["truth_xent"])
+            for m in self.memories])
+
+        statistics = [
+            F.squeeze(F.concat([F.expand_dims(ss, 1) for ss in s], axis=0))
+            if isinstance(s[0], Variable) else
+            np.array(reduce(lambda x, y: x + y.tolist(), s, []))
+            for s in statistics]
+
+        hyp_acc, truth_acc, hyp_xent, truth_xent = statistics
+
+        # Expand rewards
+        rewards = expandAlong(rewards, self.transition_mask)
```

Comment:
Yes. rewards is a (batch_size,) array, and transition_mask a (batch_size, seq_len) array, so expandAlong() populates a (batch_size, seq_len) with the rewards, and extracts only the ones when transition_mask is 1 (so rewards for skips are not in the result of expandAlong)